numpy/_core/src/multiarray/arrayobject.c

```diff
@@ -598,11 +598,12 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
         return NULL;
     }
     if (PyArray_HASFIELDS(self) && PyArray_HASFIELDS(other)) {
-        PyArray_Descr *self_descr = PyArray_DESCR(self);
-        PyArray_Descr *other_descr = PyArray_DESCR(other);
+        _PyArray_LegacyDescr *self_descr = (_PyArray_LegacyDescr *)PyArray_DESCR(self);
+        _PyArray_LegacyDescr *other_descr = (_PyArray_LegacyDescr *)PyArray_DESCR(other);
 
         /* Use promotion to decide whether the comparison is valid */
-        PyArray_Descr *promoted = PyArray_PromoteTypes(self_descr, other_descr);
+        PyArray_Descr *promoted = PyArray_PromoteTypes(
```

Comment:
Maybe do this promotion check before even defining `self_descr` and `other_descr`?

Reply:
Yeah, probably a bit nicer, changed it.